### ISSUE TITLE
VM Compiler crash on invalid Dart code (generic type)

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#34291]
An internal user on Flutter hit the following compiler crash (partially obsfucated):

> I was creating an image test file where I mistakenly included a generic type on a non-generic object. I had something like this:
> 
> ```dart
>   @override
>   Widget build(BuildContext context) {
>     return new Directionality(
>         textDirection: textDirection,
>         child: new charts.LineChart(
>           seriesList,
>           animate: false,
>           // Configures four [ChartTitle] behaviors to render titles in each
>           // chart margin.
>           behaviors: [
>             new charts.ChartTitle<num>('Top title text',
>                 subTitle: includeSubTitle ? 'Top sub-title text' : null,
>                 behaviorPosition: charts.BehaviorPosition.top,
>                 titleOutsideJustification: outsideJustification),
>           ],
>         ));
>   }
> ```
>
> `charts.ChartTitle` is not generic, so the mistaken copy-pasta `<num>` bit caused the following inscrutable error to occur. The stack trace doesn't contain any of my code, so it took a bit to work out exactly where the problem was. Now that I know the cause, I can infer that `NamedTypeError` is probably a reasonable type of error for the error. It would, however, be helpful to have a line number in my code to point the way.

The issue, here of course, is the compiler should never crash when given invalid Dart code.

```
Unhandled exception:
Crash when compiling package:flutter/material.dart,
at character offset null:
type 'QualifiedName' is not a subtype of type 'String'
#0      NamedTypeBuilder.check (package:front_end/src/fasta/builder/named_type_builder.dart:90:15)
#1      UnresolvedType.checkType (package:front_end/src/fasta/builder/unresolved_type.dart:21:31)
#2      SourceLibraryBuilder.resolveTypes (package:front_end/src/fasta/source/source_library_builder.dart:729:11)
#3      SourceLoader.resolveTypes.<anonymous closure> (package:front_end/src/fasta/source/source_loader.dart:377:36)
#4      __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.forEach (dart:collection/runtime/libcompact_hash.dart:365:8)
#5      SourceLoader.resolveTypes (package:front_end/src/fasta/source/source_loader.dart:374:14)
#6      KernelTarget.buildOutlines.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:248:18)
...
```

Assigning to @kmillikin for `area-front-end`, and /cc @a-siva.

**Comment by kmillikin** [comment]
I can't reproduce this with a simple attempt:

```dart
// in 34291.dart
import 'charts.dart' as charts;                                                                               
                                                                                                              
main() {                                                                                                      
  var c = new charts.ChartTitle<num>('text');                                                                 
} 
```
```dart
// in charts.dart
class ChartTitle {
  ChartTitle(String text);
}
```
```
$ pkg/front_end/tool/fasta compile --dump-ir --strong ~/programs/dart/34291.dart
library;                                                                                                      
import self as self;                                                                                          
import "./charts.dart" as cha;                                                                                
                                                                                                              
static method main() → dynamic {                                                                              
  cha::ChartTitle c = new cha::ChartTitle::•("text");                                                         
}                                                                                                             
library;                                                                                                      
import self as self;                                                                                          
import "dart:core" as core;                                                                                   
                                                                                                              
class ChartTitle extends core::Object {                                                                       
  constructor •(core::String text) → void                                                                     
    : super core::Object::•()                                                                                 
    ;                                                                                                         
}
```

**Comment by matanlurey** [comment]
It could be due to version skew? See internal issue is b/113349658 perhaps?

**Comment by aam** [comment]
The issue seems to be due to QualifiedName passed as a String to `templateTypeArgumentMismatch.withArguments`, so the fix seems to be simple:
```
diff --git a/pkg/front_end/lib/src/fasta/builder/named_type_builder.dart b/pkg/front_end/lib/src/fasta/builder/named_type_builder.dart
index 08b77d8840..e4945fc551 100644
--- a/pkg/front_end/lib/src/fasta/builder/named_type_builder.dart
+++ b/pkg/front_end/lib/src/fasta/builder/named_type_builder.dart
@@ -87,7 +87,7 @@ abstract class NamedTypeBuilder<T extends TypeBuilder, R> extends TypeBuilder {
           charOffset,
           fileUri,
           templateTypeArgumentMismatch.withArguments(
-              name, declaration.typeVariablesCount));
+              name.toString(), declaration.typeVariablesCount));
     }
   }
```

Only thing that is remaining is to have a test that induces this error.

**Comment by peter-ahe-google** [comment]
I'm currently reworking QualifiedName, and I'll be sure to fix this at the same time.

**Comment by peter-ahe-google** [comment]
This was fixed in 015840d31f.

This is a standalone reproduction which I'll submit as a regression test:

```
==> lib.dart <==
class A {}

==> t.dart <==
import "lib.dart" as lib;

class B {}

lib.A<B> foo() {}

main() {}
```

**Comment by peter-ahe-google** [comment]
See https://dart-review.googlesource.com/c/sdk/+/73140
